src/Sarif/PeekableStream.cs

```diff
@@ -0,0 +1,129 @@
+                return underlyingStream.Read(buffer, offset, count);
+            }
+            else if (this.cursor == this.underlyingStream.Position)
+            {
+                // Buffer cursor is "caught up" to stream position, so we are not in a "rewound" state.
+                if (this.cursor == this.rewindBuffer.Length)
+                {
+                    // A previous read has brought us to the brink of exceeding the "peekability window", and we are
+                    // now reading more.  So we abandon the buffer, and serve the read as a thin wrapper.
+                    this.rewindBuffer = null;
+                    this.cursor = int.MaxValue;
+                    return underlyingStream.Read(buffer, offset, count);
+                }
+                else
+                {
+                    // Read into our rewind buffer for "backup" and then copy into caller buffer.
+                    // Buffer position is behind Stream position.  Service read out of the buffer.
+                    int aspirationalReadCount = Math.Min(count, (int)(this.rewindBuffer.Length - cursor));
+                    int actualReadCount = underlyingStream.Read(this.rewindBuffer, cursor, aspirationalReadCount);
+                    Array.Copy(this.rewindBuffer, cursor, buffer, offset, actualReadCount);
+
+                    cursor += actualReadCount;
+
+                    return actualReadCount;
+                }
+            }
+            else if (this.cursor <= this.underlyingStream.Position)
+            {
+                // We are rewound.  Service the read out of the rewind buffer.
+                int safeReadCount = Math.Min(count, this.rewindBuffer.Length - cursor);
+                safeReadCount = Math.Min(safeReadCount, (int)(this.underlyingStream.Position - cursor));
+                Array.Copy(this.rewindBuffer, cursor, buffer, offset, safeReadCount);
+
+                cursor += safeReadCount;
+                return safeReadCount;
+            }
+            else
+            {
+                    throw new InvalidOperationException($"Peekable stream in invalid state: cursor '{cursor}' pointing past stream position '{underlyingStream.Position}'.");
+            }
+
+            // Unreachable.
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public void Rewind()
+        {
+            if (this.rewindBuffer == null)
+            {
+                throw new InvalidOperationException("This stream has read too many bytes and can not longer be rewound.");
```

Comment:
>InvalidOperationException [](http://example.com/codeflow?start=26&length=25)

is there a unit test for this condition?

Reply:
Not yet.